devices/src/virtio/vsock/unix/muxer.rs

```diff
@@ -0,0 +1,736 @@
+        };
+
+        muxer.add_listener(muxer.host_sock.as_raw_fd(), EpollListener::HostSock)?;
+        Ok(muxer)
+    }
+
+    /// Handle/dispatch an epoll event to its listener.
+    ///
+    fn handle_event(&mut self, fd: RawFd, evset: epoll::Events) {
+        debug!(
+            "vsock: muxer processing event: fd={}, evset={:?}",
+            fd, evset
+        );
+
+        match self.listener_map.get_mut(&fd) {
+            // This event needs to be forwarded to a `MuxerConnection` that is listening for it.
+            //
+            Some(EpollListener::Connection { key, evset }) => {
+                let key_copy = *key;
+                let evset_copy = *evset;
+                // The handling of this event will most probably mutate the state of the receiving
+                // conection. We'll need to check for new pending RX, event set mutation, and all that,
+                // so we're wrapping the event delivery inside those checks.
+                self.apply_conn_mutation(key_copy, |conn| {
+                    conn.notify(evset_copy);
+                });
+            }
+
+            // A new host-initiated connection is ready to be accepted.
+            //
+            Some(EpollListener::HostSock) => {
+                if self.rxq.len() == defs::MAX_CONNECTIONS {
+                    // If we're already maxed-out on connections, we'll just accept and immediately
+                    // discard this potentially new one.
+                    info!("vsock: connection limit reached; refusing new host connection");
+                    self.host_sock.accept().map(|_| 0).unwrap_or(0);
+                    return;
+                }
+                self.host_sock
+                    .accept()
+                    .map_err(Error::UnixAccept)
+                    .and_then(|(stream, _)| {
+                        stream
+                            .set_nonblocking(true)
+                            .map(|_| stream)
+                            .map_err(Error::UnixAccept)
+                    })
+                    .and_then(|stream| {
+                        // Before forwarding this connection to a listening AF_VSOCK socket on the
+                        // guest side, we need to know the destination port. We'll read that port
+                        // from a "connect" command received on this socket, so the next step is to
+                        // ask to be notified the moment we can read from it.
+                        self.add_listener(stream.as_raw_fd(), EpollListener::LocalStream(stream))
+                    })
+                    .unwrap_or_else(|err| {
+                        warn!("vsock: unable to accept local connection: {:?}", err);
+                    });
+            }
+
+            // Data is ready to be read from a host-initiated connection. That would be the "connect"
+            // command that we're expecting.
+            Some(EpollListener::LocalStream(_)) => {
```

Comment:
This is a bit not very pretty. It looks like `stream` is already in the enum variant, but that's getting ignored and we instead are using it via `self.remove_listener(fd)`. Since it seems there's not way `remove_listener` returns `None` here, can't we just use the first pattern match, and then just call `self.remove_listener(fd).unwrap()` to take out the old listener from the poll set, and get rid of the extra level of indentation caused by the `if let`?

Reply:
Ah, I remember this one - 'twas the great borrow checker battle at `handle_event()`... many a brave, if not very bright, neuron have fallen, many a curse have been uttered.

Thing is, if we capture the stream inside the match pattern, we can't call `remove_listener()`, because it will need a mutable reference to `self`, which we can't give away while holding the stream.